pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -545,6 +551,15 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 		}
 	}
 
+	// kubelet restart flow.
+	if !m.sourcesReady.AllReady() {
+		// if the container is reported running and it had a device assigned, things are already fine and we can just bail out
+		if needed == 0 && m.checkContainerAlreadyRunning(podUID, contName) {
```

Comment:
This PR doesn't address the questions about `needed == 0` in #118635. My question there was whether we could use use:
```
	if needed == 0 {
		klog.V(3).InfoS("no devices needed, nothing to do", "deviceNumber", needed, "resourceName", resource, "podUID", string(podUID), "containerName", contName)
		// No change, no work.
		return nil, nil
	}
```
As the first check in this function.

Reply:
You are totally right, I should have mentioned this more explicitely. The reason why I didn't address your comment is I'm still evaluating the flow here. I tend to agree we can simplify, but I want to run a bit more tests before to change the code (or explain why I believe better not to change).